### PR TITLE
Particle system - Part 1

### PR DESCRIPTION
**Pull request by caguero**
It depends on https://github.com/ignitionrobotics/ign-msgs/pull/127 .

The goal is to integrate a particle system that can use the particle emitters available in Ignition Rendering.

This patch focuses on the creation of a new `ParticleEmitter` component.

### CODE REVIEW

File: include/ignition/gazebo/components/ParticleEmitter.hh
Comment (chapulina) [inline include/ignition/gazebo/components/ParticleEmitter.hh:50]: So far we haven't been creating entirely new classes to hold component data because there's usually already a class in some other library that can be used here and we don't duplicate information. I'd be curious to see the rest of the implementation to understand how this will fit with the rest of the code. Are we creating a message to hold this data? How about an SDF class? Another thing to keep in mind is that this header-only struct can't be extended with new fields without breaking ABI.

All this questioning is related to #494 :upside_down_face:

Comment (caguero) [reply include/ignition/gazebo/components/ParticleEmitter.hh:50]: @iche033 and I brainstormed about it when we were thinking about the particle integration and we thought this was the best alternative. I followed the same approach that we did with `LogicalAudio` in pull request #401 . 

The plan is to use an Ignition Message to store this data (there's already a [branch](https://github.com/ignitionrobotics/ign-msgs/blob/particle_emiter/proto/ignition/msgs/particle_emiter.proto) for it) and use a component that can be serialized to reach the clients. Then, in the rendering thread, use the Ignition Rendering API to create the particle emitter based on the request.

I wasn't planning to create an SDF class. Our idea is th

Comment (iche033) [reply include/ignition/gazebo/components/ParticleEmitter.hh:50]: We discussed whether or not we want to commit to adding this feature to the official sdf spec and create the DOM classes. But in the end, we thought adding to ign gazebo first seems like a less disruptive option that we could do in Dome for subt. 

I'm not opposed to adding to sdf spec. This could be a chance for something like `<visual><geometry><particles>` but we may want to include the sdf folks in the discussion / design

Comment (chapulina) [reply include/ignition/gazebo/components/ParticleEmitter.hh:50]: >  I followed the same approach that we did with LogicalAudio

Oh ok I hadn't realized that was the approach there. My suggestion would be to use `ignition::msgs::ParticleEmitter` as the data for the component. That should be more efficient to serialize, and we don't need to worry about keeping the message and the component in sync. It looks like the new class isn't doing more than the message would, so I think there isn't a need for it.

> adding this feature to the official sdf spec

Yeah I think it makes sense to configure this feature through the `<plugin>` tag for now.

Comment (caguero) [reply include/ignition/gazebo/components/ParticleEmitter.hh]: Thanks for the suggestion, it looks better in the code to me. How about now? Is this what you were thinking?

Comment (chapulina) [reply include/ignition/gazebo/components/ParticleEmitter.hh]: Much closer! Now that all the data is contained in the message, I think we could get rid of the `ignition::gazebo::particles::Emitter` class and just use the message directly in the component, right? Even if we think we may need to extend the class later, I think it would break ABI just like extending the message, so we can just extend the message.

When you use the message as the data, you don't need the custom operators, the custom serializer... It should be possible to define the component in one line like this:

https://github.com/ignitionrobotics/ign-gazebo/blob/a830b443989877b1e83170390f483b6c02952b25/include/ignition/gazebo/components/ExternalWorldWrenchCmd.hh#L41-L43

(also, let's not forget to bump the required `ign-msgs` version before merging this PR)

Comment (caguero) [reply include/ignition/gazebo/components/ParticleEmitter.hh]: Done in 4cbc9dd! How about now?

**Review by chapulina — APPROVED** [review]
Thanks for iterating!

**Comment by chapulina** [comment]
@caguero , would you prefer merging these PRs as they are approved, or should we wait until all the functionality is working and approved before merging all tutorials? Note that this is targeting a stable release branch, and whatever is merged in an intermediary state may be released at any point.

**Comment by caguero** [comment]
> @caguero , would you prefer merging these PRs as they are approved, or should we wait until all the functionality is working and approved before merging all tutorials? Note that this is targeting a stable release branch, and whatever is merged in an intermediary state may be released at any point.

We can review part1, part2 and part3 and merge them all at once when we're happy with it. These three PRs offer the basic integration. Part4 (coming up) is more independent.

**Comment by chapulina** [comment]
> We can review part1, part2 and part3 and merge them all at once when we're happy with it.

Got it. I converted this to draft so that we don't merge it by mistake.

**Comment by adlarkin** [comment]
I'm noticing the following error in CI:

```
  /github/workspace/include/ignition/gazebo/components/ParticleEmitter.hh:20:10: fatal error: ignition/msgs/particle_emitter.pb.h: No such file or directory
   #include <ignition/msgs/particle_emitter.pb.h>
```

I believe that we need to make an `ign-msgs6` release. cc @chapulina @iche033

**Comment by chapulina** [comment]
> ign-msgs6 release.

I'm fine with a release, my only request is that we make sure the current particle message has all we need, because this will go into a stable release, and we won't be able to tweak any fields afterwards. So maybe it's best to wait until all parts are out beforehand. But if @iche033 thinks the implementation is already settled, than go for it.

**Comment by adlarkin** [comment]
> my only request is that we make sure the current particle message has all we need, because this will go into a stable release, and we won't be able to tweak any fields afterwards.

Good point. I am still reviewing the other particle system PRs, so I will make a release once we know all of the other PRs are good to go :+1:

**Comment by iche033** [comment]
yeah let's hold off until we have everything reviewed and ready to go. There's still some work left that need to be on on particles this week.

**Comment by adlarkin** [comment]
@iche033 there was a merge conflict in https://github.com/ignitionrobotics/ign-gazebo/pull/516/commits/5885e45de53dea0833e8dea440e83e154eef2fc9 that I resolved. It looks like #482 added some code to `RenderUtil::UpdateECM` to update lights, so I kept that addition along with the additions we made to `RenderUtil::UpdateECM`.

File: src/systems/particle_emitter/ParticleEmitter.cc
Comment (adlarkin) [inline src/systems/particle_emitter/ParticleEmitter.cc:151]: We should probably leave a note that the `name` field in the emitter message never seems to be used past this point (at least, I don't see the message's `name` being used anywhere else in this PR).

File: test/integration/particle_emitter.cc
Comment (adlarkin) [inline test/integration/particle_emitter.cc:106]: I wasn't sure how to load/test for materials, so I left this check out. If anyone wants to add this check in before merging, that'd be great, but if not, we can do this later.

File: test/integration/particle_emitter.cc
Comment (adlarkin) [inline test/integration/particle_emitter.cc:142]: I wasn't sure how to load/test for materials, so I left this check out. If anyone wants to add this check in before merging, that'd be great, but if not, we can do this later.

File: test/integration/particle_emitter.cc
Comment (adlarkin) [inline test/integration/particle_emitter.cc:170]: Do we need this? I don't think we use this in other integration tests.

**Review by adlarkin — APPROVED** [review]
A few minor comments that would be nice to address before merging, but I'm fine with merging this PR as-is.

File: src/systems/particle_emitter/ParticleEmitter.cc
Comment (nkoenig) [reply src/systems/particle_emitter/ParticleEmitter.cc:151]: The name is also used in the `component` and topic name. I'll leave this code as-is.

File: test/integration/particle_emitter.cc
Comment (nkoenig) [reply test/integration/particle_emitter.cc:106]: Leaving it for later.

File: test/integration/particle_emitter.cc
Comment (nkoenig) [reply test/integration/particle_emitter.cc:142]: Leaving it for later.

File: test/integration/particle_emitter.cc
Comment (nkoenig) [reply test/integration/particle_emitter.cc]: It's not required. I removed it in 8a1db3d3012bfdcfb673c266c998952d77475716

**Pull request merged**
